File: data-stack/consolidated-data-stack/lib/stacks/msk-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as msk from 'aws-cdk-lib/aws-msk';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as cr from 'aws-cdk-lib/custom-resources';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';
import { Config } from '../config';
import * as path from 'path';

export interface MskStackProps extends cdk.StackProps {
  vpc: ec2.IVpc;
  mskSecurityGroup: ec2.SecurityGroup;
}

export class MskStack extends cdk.Stack {
  public readonly cluster: msk.CfnCluster;
  public readonly bootstrapServers: string;
  public readonly logGroup: logs.LogGroup;

  constructor(scope: Construct, id: string, props: MskStackProps) {
    super(scope, id, props);

    // S3 bucket for MSK logs
    const logsBucket = new s3.Bucket(this, 'MskLogsBucket', {
      bucketName: `${Config.s3.logsBucketName}-${this.account}-${this.region}`,
      encryption: s3.BucketEncryption.S3_MANAGED,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    // CloudWatch log group for MSK
    this.logGroup = new logs.LogGroup(this, 'MskLogGroup', {
      logGroupName: `/aws/msk/${Config.msk.clusterName}`,
      retention: logs.RetentionDays.ONE_WEEK,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // Get private subnet IDs
    const privateSubnets = props.vpc.selectSubnets({
      subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
    }).subnetIds;

    // MSK Cluster configuration
    // Shared IAM role for all MSK custom resources to avoid policy race conditions
    const mskCustomResourceRole = new iam.Role(this, 'MskCustomResourceRole', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
      ],
    });
    mskCustomResourceRole.addToPolicy(new iam.PolicyStatement({
      actions: [
        'kafka:UpdateConnectivity',
        'kafka:DescribeCluster',
        'kafka:DescribeClusterV2',
        'kafka:PutClusterPolicy',
        'kafka:GetClusterPolicy',
        'kafka:CreateConfiguration',
        'kafka:ListConfigurations',
        'kafka:DescribeConfiguration',
        'kafka:UpdateClusterConfiguration',
        'kafka:GetBootstrapBrokers',
      ],
      resources: ['*'],
    }));

    this.cluster = new msk.CfnCluster(this, 'MskCluster', {
      clusterName: Config.msk.clusterName,
      kafkaVersion: Config.msk.kafkaVersion,
      numberOfBrokerNodes: Config.msk.brokerCount,
      brokerNodeGroupInfo: {
        instanceType: Config.msk.brokerInstanceType,
        clientSubnets: privateSubnets,
        securityGroups: [props.mskSecurityGroup.securityGroupId],
        storageInfo: {
          ebsStorageInfo: {
            volumeSize: Config.msk.ebsVolumeSize,
          },
        },
      },
      clientAuthentication: {
        sasl: {
          iam: { enabled: true },
        },
      },
      encryptionInfo: {
        encryptionInTransit: {
          clientBroker: 'TLS',
          inCluster: true,
        },
      },
      enhancedMonitoring: 'PER_TOPIC_PER_BROKER',
      loggingInfo: {
        brokerLogs: {
          cloudWatchLogs: {
            enabled: true,
            logGroup: this.logGroup.logGroupName,
          },
          s3: {
            enabled: true,
            bucket: logsBucket.bucketName,
            prefix: 'msk-logs/',
          },
        },
      },
    });

    // Custom resource to enable Multi-VPC Private Connectivity for Firehose integration
    // This must be done via API call, not CloudFormation property
    const enableVpcConnectivity = new cr.AwsCustomResource(this, 'EnableVpcConnectivity', {
      onCreate: {
        service: 'Kafka',
        action: 'updateConnectivity',
        parameters: {
          ClusterArn: this.cluster.attrArn,
          CurrentVersion: this.cluster.attrCurrentVersion,
          ConnectivityInfo: {
            VpcConnectivity: {
              ClientAuthentication: {
                Sasl: {
                  Iam: { Enabled: true },
                },
              },
            },
          },
        },
        physicalResourceId: cr.PhysicalResourceId.of('EnableVpcConnectivity'),
      },
      role: mskCustomResourceRole,
      policy: cr.AwsCustomResourcePolicy.fromSdkCalls({ resources: cr.AwsCustomResourcePolicy.ANY_RESOURCE }),
    });
    enableVpcConnectivity.node.addDependency(this.cluster);

    // Add cluster policy to allow Firehose to create VPC connections
    const clusterPolicy = {
      Version: '2012-10-17',
      Statement: [
        {
          Effect: 'Allow',
          Principal: {
            Service: 'firehose.amazonaws.com',
          },
          Action: [
            'kafka:CreateVpcConnection',
            'kafka:GetBootstrapBrokers',
            'kafka:DescribeCluster',
            'kafka:DescribeClusterV2',
          ],
          Resource: this.cluster.attrArn,
        },
      ],
    };

    const putClusterPolicy = new cr.AwsCustomResource(this, 'PutClusterPolicy', {
      onCreate: {
        service: 'Kafka',
        action: 'putClusterPolicy',
        parameters: {
          ClusterArn: this.cluster.attrArn,
          Policy: JSON.stringify(clusterPolicy),
        },
        physicalResourceId: cr.PhysicalResourceId.of('ClusterPolicy'),
      },
      onUpdate: {
        service: 'Kafka',
        action: 'putClusterPolicy',
        parameters: {
          ClusterArn: this.cluster.attrArn,
          Policy: JSON.stringify(clusterPolicy),
        },
        physicalResourceId: cr.PhysicalResourceId.of('ClusterPolicy'),
      },
      role: mskCustomResourceRole,
      policy: cr.AwsCustomResourcePolicy.fromSdkCalls({ resources: cr.AwsCustomResourcePolicy.ANY_RESOURCE }),
    });
    putClusterPolicy.node.addDependency(enableVpcConnectivity);

    // MSK Configuration with auto-topic creation enabled
    // Uses Lambda-backed custom resource for idempotent creation (handles "already exists" case)
    const configName = `${Config.msk.clusterName}-auto-create-topics`;
    const serverProperties = [
      'auto.create.topics.enable=true',
      'default.replication.factor=3',
      'min.insync.replicas=2',
      'num.partitions=3',
    ].join('\n');

    const mskConfigHandler = new lambda.Function(this, 'MskConfigHandler', {
      runtime: lambda.Runtime.PYTHON_3_11,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../../lambda/msk-config')),
      timeout: cdk.Duration.minutes(2),
      role: mskCustomResourceRole,
    });

    const mskConfigProvider = new cr.Provider(this, 'MskConfigProvider', {
      onEventHandler: mskConfigHandler,
    });

    const mskConfig = new cdk.CustomResource(this, 'MskConfiguration', {
      serviceToken: mskConfigProvider.serviceToken,
      properties: {
        ConfigurationName: configName,
        KafkaVersions: [Config.msk.kafkaVersion],
        ServerProperties: serverProperties,
      },
    });

    // Get current cluster version for update
    // Note: We need to fetch the current version right before updating
    // since it changes after each cluster operation
    const getClusterVersion = new cr.AwsCustomResource(this, 'GetClusterVersion', {
      onCreate: {
        service: 'Kafka',
        action: 'describeClusterV2',
        parameters: {
          ClusterArn: this.cluster.attrArn,
        },
        physicalResourceId: cr.PhysicalResourceId.of('GetClusterVersion'),
      },
      role: mskCustomResourceRole,
      policy: cr.AwsCustomResourcePolicy.fromSdkCalls({ resources: cr.AwsCustomResourcePolicy.ANY_RESOURCE }),
    });
    getClusterVersion.node.addDependency(putClusterPolicy);
    getClusterVersion.node.addDependency(mskConfig);

    // Update cluster to use the MSK configuration
    const updateClusterConfig = new cr.AwsCustomResource(this, 'UpdateClusterConfiguration', {
      onCreate: {
        service: 'Kafka',
        action: 'updateClusterConfiguration',
        parameters: {
          ClusterArn: this.cluster.attrArn,
          CurrentVersion: getClusterVersion.getResponseField('ClusterInfo.CurrentVersion'),
          ConfigurationInfo: {
            Arn: mskConfig.getAttString('Arn'),
            Revision: 1,
          },
        },
        physicalResourceId: cr.PhysicalResourceId.of('UpdateClusterConfiguration'),
      },
      role: mskCustomResourceRole,
      policy: cr.AwsCustomResourcePolicy.fromSdkCalls({ resources: cr.AwsCustomResourcePolicy.ANY_RESOURCE }),
    });
    updateClusterConfig.node.addDependency(getClusterVersion);

    // Custom resource to get bootstrap servers
    const getBootstrapServers = new cr.AwsCustomResource(this, 'GetBootstrapServers', {
      onCreate: {
        service: 'Kafka',
        action: 'getBootstrapBrokers',
        parameters: {
          ClusterArn: this.cluster.attrArn,
        },
        physicalResourceId: cr.PhysicalResourceId.of('BootstrapServers'),
      },
      role: mskCustomResourceRole,
      policy: cr.AwsCustomResourcePolicy.fromSdkCalls({ resources: cr.AwsCustomResourcePolicy.ANY_RESOURCE }),
    });
    getBootstrapServers.node.addDependency(this.cluster);

    this.bootstrapServers = getBootstrapServers.getResponseField('BootstrapBrokerStringSaslIam');

    // Outputs
    new cdk.CfnOutput(this, 'MskClusterArn', { value: this.cluster.attrArn });
    new cdk.CfnOutput(this, 'BootstrapServers', { value: this.bootstrapServers });
  }
}

File: data-stack/to_be_deleted/ibc2025-data-gen-acme-video-telemetry-synthetic/cdk/stacks/msk_stack.py
"""
ACME Telemetry Pipeline - MSK Stack
Optional stack for creating Amazon MSK cluster if not using existing cluster
"""

from aws_cdk import (
    Stack,
    RemovalPolicy,
    aws_msk as msk,
    aws_ec2 as ec2,
    aws_iam as iam,
    aws_logs as logs,
    CfnOutput
)
from constructs import Construct

class MSKStack(Stack):
    """Optional MSK cluster stack for ACME telemetry pipeline"""
    
    def __init__(self, scope: Construct, construct_id: str, 
                 vpc: ec2.IVpc, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)
        
        self.vpc = vpc
        
        # Security group for MSK cluster
        self.msk_security_group = ec2.SecurityGroup(
            self, "MSKSecurityGroup",
            vpc=self.vpc,
            description="Security group for MSK cluster",
            allow_all_outbound=True
        )
        
        # Allow internal communication
        self.msk_security_group.add_ingress_rule(
            peer=self.msk_security_group,
            connection=ec2.Port.all_tcp(),
            description="Allow MSK internal communication"
        )
        
        # Allow Lambda access on IAM auth port
        self.msk_security_group.add_ingress_rule(
            peer=ec2.Peer.ipv4(self.vpc.vpc_cidr_block),
            connection=ec2.Port.tcp(9098),
            description="MSK IAM authentication"
        )
        
        # Allow plaintext for internal communication (optional)
        self.msk_security_group.add_ingress_rule(
            peer=ec2.Peer.ipv4(self.vpc.vpc_cidr_block),
            connection=ec2.Port.tcp(9092),
            description="MSK plaintext"
        )
        
        # CloudWatch log group for MSK
        self.log_group = logs.LogGroup(
            self, "MSKLogGroup",
            log_group_name="/aws/msk/acme-telemetry-cluster",
            retention=logs.RetentionDays.ONE_WEEK,
            removal_policy=RemovalPolicy.DESTROY
        )
        
        # MSK Configuration
        msk_config = msk.CfnConfiguration(
            self, "MSKConfiguration",
            name="acme-telemetry-config",
            kafka_versions_list=["2.8.1"],
            server_properties="""
auto.create.topics.enable=true
default.replication.factor=3
min.insync.replicas=2
num.partitions=20
log.retention.hours=168
compression.type=gzip
"""
        )
        
        # Create MSK Cluster
        self.cluster = msk.CfnCluster(
            self, "MSKCluster",
            cluster_name="acme-telemetry-msk",
            kafka_version="2.8.1",
            number_of_broker_nodes=3,
            broker_node_group_info=msk.CfnCluster.BrokerNodeGroupInfoProperty(
                instance_type="kafka.m5.large",
                client_subnets=[subnet.subnet_id for subnet in self.vpc.private_subnets[:3]],
                security_groups=[self.msk_security_group.security_group_id],
                storage_info=msk.CfnCluster.StorageInfoProperty(
                    ebs_storage_info=msk.CfnCluster.EBSStorageInfoProperty(
                        volume_size=100,
                        provisioned_throughput=msk.CfnCluster.ProvisionedThroughputProperty(
                            enabled=False
                        )
                    )
                ),
                connectivity_info=msk.CfnCluster.ConnectivityInfoProperty(
                    public_access=msk.CfnCluster.PublicAccessProperty(
                        type="DISABLED"
                    ),
                    vpc_connectivity=msk.CfnCluster.VpcConnectivityProperty(
                        client_authentication=msk.CfnCluster.VpcConnectivityClientAuthenticationProperty(
                            sasl=msk.CfnCluster.VpcConnectivitySaslProperty(
                                iam=msk.CfnCluster.VpcConnectivityIamProperty(
                                    enabled=True
                                )
                            )
                        )
                    )
                )
            ),
            client_authentication=msk.CfnCluster.ClientAuthenticationProperty(
                sasl=msk.CfnCluster.SaslProperty(
                    iam=msk.CfnCluster.IamProperty(
                        enabled=True
                    )
                )
            ),
            configuration_info=msk.CfnCluster.ConfigurationInfoProperty(
                arn=msk_config.ref,
                revision=1
            ),
            encryption_info=msk.CfnCluster.EncryptionInfoProperty(
                encryption_in_transit=msk.CfnCluster.EncryptionInTransitProperty(
                    client_broker="TLS",
                    in_cluster=True
                )
            ),
            enhanced_monitoring="PER_TOPIC_PER_BROKER",
            logging_info=msk.CfnCluster.LoggingInfoProperty(
                broker_logs=msk.CfnCluster.BrokerLogsProperty(
                    cloud_watch_logs=msk.CfnCluster.CloudWatchLogsProperty(
                        enabled=True,
                        log_group=self.log_group.log_group_name
                    )
                )
            ),
            tags={
                "Name": "acme-telemetry-msk",
                "Project": "ACME-Telemetry",
                "Environment": "Production"
            }
        )
        
        # Ensure log group is created before cluster
        self.cluster.node.add_dependency(self.log_group)
        
        # Create cluster policy for Firehose access
        cluster_policy = iam.PolicyDocument(
            statements=[
                iam.PolicyStatement(
                    effect=iam.Effect.ALLOW,
                    principals=[iam.ServicePrincipal("firehose.amazonaws.com")],
                    actions=[
                        "kafka:CreateVpcConnection",
                        "kafka:GetBootstrapBrokers",
                        "kafka:DescribeCluster",
                        "kafka:DescribeClusterV2",
                        "kafka-cluster:Connect",
                        "kafka-cluster:DescribeCluster",
                        "kafka-cluster:ReadData",
                        "kafka-cluster:DescribeGroup",
                        "kafka-cluster:AlterGroup",
                        "kafka-cluster:DescribeTopic"
                    ],
                    resources=["*"],
                    conditions={
                        "StringEquals": {
                            "aws:SourceAccount": self.account
                        }
                    }
                )
            ]
        )
        
        # Note: Cluster policy would need to be applied via Custom Resource or CLI
        CfnOutput(
            self, "ClusterPolicyCommand",
            value=f"aws kafka put-cluster-policy --cluster-arn {self.cluster.attr_arn} --policy '{cluster_policy.to_json()}'",
            description="Run this command to apply the cluster policy"
        )
        
        # Outputs
        CfnOutput(
            self, "MSKClusterArn",
            value=self.cluster.attr_arn,
            description="ARN of the MSK cluster",
            export_name="MSKClusterArn"
        )
        
        CfnOutput(
            self, "MSKSecurityGroupId",
            value=self.msk_security_group.security_group_id,
            description="Security group ID for MSK cluster"
        )